libyara/exec.c

```diff
@@ -194,21 +194,117 @@ static const uint8_t* jmp_if(
 }
 
 
+static int iter_array_next(
+    YR_ITERATOR* self,
+    YR_VALUE_STACK* stack)
+{
+  YR_OBJECT* obj;
+
+  // Check that there's two available slots in the stack, one for the next
+  // item returned by the iterator and another one for the boolean that
+  // indicates if there are more items.
+  if (stack->sp + 1 >= stack->capacity)
+    return ERROR_EXEC_STACK_OVERFLOW;
+
+  if (self->array_it.index < yr_object_array_length(self->array_it.array))
+  {
+    // Push the false value that indicates that the iterator is not exhausted.
+    stack->items[stack->sp++].i = 0;
+
+    obj = yr_object_array_get_item(
+        self->array_it.array, 0, self->array_it.index);
+
+    if (obj != NULL)
+      stack->items[stack->sp++].o = obj;
+    else
+      stack->items[stack->sp++].i = UNDEFINED;
```

Comment:
Rather than push UNDEFINED can't we just skip this item? I'm thinking of a scenario where someone has done something like this in a module:

`
set_integer(1, pe->object, "foo[%i]", 0);
set_integer(2, pe->object, "foo[%i]", 100);
`

In this case wouldn't we be pushing a lot of UNDEFINED values before we get to the next assigned value in the array? Wouldn't it be nicer to only return objects if they are there and skip over the unassigned array values?

Reply:
Actually that was my first idea, but I found out that it wouldn't work as expected. 

Suppose you have an array of integers with 3 items where the first two are undefined, like [undefined, undefined, 10]. If we simply skip the first two items that would be undistinguishable from an array with a single item, the loop would be executed only once, and an expression like this one would be true:

```
for all i in array : ( i == 10)
```

However this expression shouldn't be true because the equality is not true for all items in the array as there are two of them that are not defined.

Reply:
That makes sense. The ways I see around this involve some rather drastic, and possibly backwards-compatability breaking, changes. In particular I'm thinking we can do something like `set_integer(1, pe->object, "foo")` and have it act more like an append function, where you don't get to specify the index. This would at least ensure the values that are set are contiguous, though it doesn't address the fact that you can still set undefined. As I think about it more, I think your approach is the best - it sucks that we can iterate over an arbitrary number of UNDEFINED values but it is the best option.

Thanks for the clarification here!